compdb: minor fixes to the COMPDB lump

fabiangreffrath opened the pull request.

fabiangreffrath commented:
This will make it easier to combine similar rules for multiple maps (e.g. from the same PWAD), as suggested by @elf-alchemist in #2121.

rfomin commented:
> This will make it easier to combine similar rules for multiple maps (e.g. from the same PWAD), as suggested by @elf-alchemist in #2121.

We can combine most of the records because they have the same rules (`"options": [ {"name": "comp_vile", "value": 1} ]`), but it will be less readable. I don't think it's worth it.

fabiangreffrath commented:
I'm not going to combine the entries in our current `compdb.lmp` database. But this change will make it easier to create a custom `compdb.lmp` database specifically for PWADs like TimeOfDeath's Eternal Slumber Party, as outlined in #2121.

fabiangreffrath commented:
BTW, we have a `.clang-format` file for a consistent formatting of new C code files that we added. Should we do the same for our own JSON files that we added and e.g. pipe them all through `json_pp`?

rfomin commented:
COMPDB records are already stored in the `levels` array, so we might want to rename or reorganize it.

> But this change will make it easier to create a custom compdb.lmp database specifically for PWADs like TimeOfDeath's Eternal Slumber Party, as outlined in https://github.com/fabiangreffrath/woof/issues/2121.

The idea of COMPDB is to play in MBF21 complevel (which is more comfortable with various Boom issues fixed) with some compatibility flags enabled. Often mappers have made a level with Boom compatibility, but it is still perfectly playable in MBF21. For "Eternal Slumber Party" we don't know if all "limit removing" levels are unplayable, DoomWiki only mentions [MAP21](https://doomwiki.org/wiki/Eternal_Slumber_Party).

My point is that we shouldn't force mass complevel changes on many maps at once.

rfomin commented:
> BTW, we have a `.clang-format` file for a consistent formatting of new C code files that we added. Should we do the same for our own JSON files that we added and e.g. pipe them all through `json_pp`?

Never heard of `json_pp`. `.clang format` for JSON is [schema](https://json-schema.org/), most editors support it. It looks overcomplicated though.

fabiangreffrath commented:
> COMPDB records are already stored in the `levels` array, so we might want to rename or reorganize it.

It would still be "levels" after all. The only difference is that we would treat them in sets instead of strictly one-by-one.

> The idea of COMPDB is to play in MBF21 complevel (which is more comfortable with various Boom issues fixed) with some compatibility flags enabled. Often mappers have made a level with Boom compatibility, but it is still perfectly playable in MBF21. For "Eternal Slumber Party" we don't know if all "limit removing" levels are unplayable, DoomWiki only mentions [MAP21](https://doomwiki.org/wiki/Eternal_Slumber_Party).
> 
> My point is that we shouldn't force mass complevel changes on many maps at once.

I wouldn't want to add the "Eternal Slumber Party" entries to our own database. Don't get me wrong, I don't want to change anything beside this PR. I just want to create the *possibility* for others to create their own compatibility databases more comfortably.

> Never heard of `json_pp`. `.clang-format` for JSON is [schema](https://json-schema.org/), most editors support it. It looks overcomplicated though.

`json_pp` is part of the Perl core installation. It's not complicated and mostly a "do once" operation: `json_pp < in.json > out.json`.

rfomin reviewed (approved):
Looks good.

fabiangreffrath commented:
> Looks good.

Thank you! I'll apply some more tests at home before I merge this, though.

rfomin commented:
> `json_pp` is part of the Perl core installation. It's not complicated and mostly a "do once" operation: `json_pp < in.json > out.json`.

Perl is broken on Windows (popular distribution adds it's own GCC distribution to PATH). Do you want to format JSON files in GitHub action or just add some config for `json_pp`? If the latter, no problem.

fabiangreffrath commented:
> Do you want to format JSON files in GitHub action or just add some config for `json_pp`? If the latter, no problem.

Neither. My idea was to pipe the JSON format lumps through the pretty-printer once and commit the changes here.

A more potable alternative would be this:

```python
import sys
import json

js = json.load(sys.stdin)
json.dump(js, sys.stdout, indent=2)
```

rfomin commented:
> Neither. My idea was to pipe the JSON format lumps through the pretty-printer once and commit the changes here.

Okay, no pronblem.
